code/game/objects/items/devices/lighting/toggleable/flashlight.dm

```diff
@@ -3,33 +3,225 @@
+
+/obj/item/device/lighting/toggleable/flashlight/proc/place_lightspot(var/turf/T, var/angle = null, var/hitObstacle = FALSE)
+	if (light_spot && on && !T.is_space())
+		light_spot.forceMove(T)
+		light_spot.icon_state = "nothing"
+		light_spot.transform = initial(light_spot.transform)
+		if (lightSpotPlaceable(T) && !hitObstacle)
+			var/distance = get_dist(get_turf(src),T)
+			switch(distance)
+				if (1)
+					light_spot.icon_state = "lightspot_vclose"
+				if (2)
+					light_spot.icon_state = "lightspot_close"
+				if (3)
+					light_spot.icon_state = "lightspot_medium"
+				if (4)
+					light_spot.icon_state = "lightspot_far"
+
+		if(angle)
+			light_spot.transform = turn(light_spot.transform, angle)
+		else
+			switch(light_direction)	//icon pointing north by default
+				if(SOUTH)
+					light_spot.transform = turn(light_spot.transform, 180)
+				if(EAST)
+					light_spot.transform = turn(light_spot.transform, 90)
+				if(WEST)
+					light_spot.transform = turn(light_spot.transform, -90)
+
+/obj/item/device/lighting/toggleable/flashlight/proc/lightSpotPassable(var/turf/T)
+	if (is_opaque(T))
+		return FALSE
+	return TRUE
+
+/obj/item/device/lighting/toggleable/flashlight/proc/lightSpotPlaceable(var/turf/T)	//check if we can place icon there, light will be still applied
+	if(T == get_turf(src) || !isfloor(T))
+		return FALSE
+	for(var/obj/O in T)
+		if(istype(O, /obj/structure/window))
+			return FALSE
+	return TRUE
+
+/obj/item/device/lighting/toggleable/flashlight/moved(mob/user, old_loc)
+	spot_locked = FALSE
+	calculate_dir(old_loc)
+
+/obj/item/device/lighting/toggleable/flashlight/entered_with_container()
+	spot_locked = FALSE
+	calculate_dir()
+
+/obj/item/device/lighting/toggleable/flashlight/container_dir_changed(new_dir)
+	spot_locked = FALSE
+	set_dir(new_dir)
+
+/obj/item/device/lighting/toggleable/flashlight/pickup(mob/user)
+	..()
+	calculate_dir()
+	dir = WEST
+
+/obj/item/device/lighting/toggleable/flashlight/dropped(mob/user as mob)
+	if(light_direction)
+		set_dir(light_direction)
+
+/obj/item/device/lighting/toggleable/flashlight/afterattack(atom/A, mob/user)
+	var/turf/T = get_turf(A)
+	if(can_see(user,T) && light_spot_range >= get_dist(get_turf(src),T))
+		if (!lightSpotPassable(T))
+			T = get_step_towards(T,get_turf(src))
+			if(!lightSpotPassable(T))
+				return
+		spot_locked = TRUE
+		light_direction = get_dir(src,T)
+		place_lightspot(T,Get_Angle(get_turf(src),T))
+
 /obj/item/device/lighting/toggleable/flashlight/turn_on(mob/user)
 	if(!cell || !cell.check_charge(tick_cost))
 		playsound(loc, 'sound/machines/button.ogg', 50, 1)
 		user << SPAN_WARNING("[src] battery is dead or missing.")
 		return FALSE
 	. = ..()
+	set_light(2,radiance_power)
+	light_spot = new(get_turf(src),light_spot_radius, light_spot_power)
+	light_spot.icon = 'icons/effects/64x64.dmi'
+	light_spot.pixel_x = -16
+	light_spot.pixel_y = -16
+	light_spot.layer = ABOVE_OBJ_LAYER
+	if (cell.get_charge_percent() <= 15)
+		apply_power_deficiency()
```

Comment:
Power deficiency? What's this mechanic about?

Reply:
While cell in flashlight dies, it will affect flashlight luminosity, it will warn player that flashlight is about to die(its not funny to be blind like a mole in the dark). We can move this proc and checks somewhere else like /obj/item/device so all devices could potentially have defects based on charge left.